pkg/bwagreement/server.go

```diff
@@ -110,75 +107,66 @@ func (dbm *DBManager) GetBandwidthAllocations(ctx context.Context) (rows []*dbx.
 }
 
 // BandwidthAgreements receives and stores bandwidth agreements from storage nodes
-func (s *Server) BandwidthAgreements(stream pb.Bandwidth_BandwidthAgreementsServer) (err error) {
-	ctx := stream.Context()
+func (s *Server) BandwidthAgreements(ctx context.Context, agreement *pb.RenterBandwidthAllocation) (reply *pb.AgreementsSummary, err error) {
 	defer mon.Task()(&ctx)(&err)
 	defer s.dbm.locked()()
 
-	ch := make(chan *pb.RenterBandwidthAllocation, 1)
-	errch := make(chan error, 1)
-	go func() {
-		for {
-			msg, err := stream.Recv()
-			if err != nil {
-				s.logger.Error("Grpc Receive Error", zap.Error(err))
-				errch <- err
-				return
-			}
-			ch <- msg
-		}
-	}()
-
-	for {
-		select {
-		case err := <-errch:
-			return err
-		case <-ctx.Done():
-			return nil
-		case agreement := <-ch:
-			if err = s.verifySignature(ctx, agreement); err != nil {
-				return err
-			}
-			_, err = s.dbm.Create(ctx, agreement)
-			if err != nil {
-				s.logger.Error("DB entry creation Error", zap.Error(err))
-				return err
-			}
-		}
+	s.logger.Debug("Received Agreement...")
+
+	reply = &pb.AgreementsSummary{
+		Status: pb.AgreementsSummary_FAIL,
+	}
+
+	if err = s.verifySignature(ctx, agreement); err != nil {
+		return reply, err
 	}
 
+	_, err = s.dbm.Create(ctx, agreement)
+	if err != nil {
+		s.logger.Error("DB entry creation Error", zap.Error(err))
+		return reply, err
+	}
+
+	reply.Status = pb.AgreementsSummary_OK
+
+	s.logger.Debug("Stored Agreement...")
+
+	return reply, nil
 }
 
 func (s *Server) verifySignature(ctx context.Context, ba *pb.RenterBandwidthAllocation) error {
 	// TODO(security): detect replay attacks
-	pi, err := provider.PeerIdentityFromContext(ctx)
+
+	//Deserealize RenterBandwidthAllocation.GetData() so we can get public key
+	rbad := &pb.RenterBandwidthAllocation_Data{}
```

Comment:
Is there a reason we are not using the pub-key that is passed along with TLS connection?

Reply:
We need the uplink's public key. These messages are being sent from the Storage nodes. The storage nodes need to hold onto the public key used for each allocation